pkg/contexts/ocm/pubsub/providers/ocireg/provider.go

```diff
@@ -0,0 +1,178 @@
+package ocireg
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+
+	"github.com/mandelsoft/goutils/errors"
+
+	"github.com/open-component-model/ocm/pkg/blobaccess/blobaccess"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/cpi"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/cpi/repocpi"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/pubsub"
+	compound2 "github.com/open-component-model/ocm/pkg/contexts/ocm/pubsub/types/compound"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/repositories/genericocireg"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/repositories/genericocireg/componentmapping"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/repositories/ocireg"
+)
+
+const (
+	ConfigMimeType     = "application/vnd.ocm.software.repository.config.v1+json"
+	PubSubLayerMimeTye = "application/vnd.ocm.software.repository.config.pubsub.v1+json"
+)
+
+const META = "meta"
+
+func init() {
+	pubsub.RegisterProvider(ocireg.Type, &Provider{})
+}
+
+type Provider struct{}
+
+var _ pubsub.Provider = (*Provider)(nil)
+
+func (p *Provider) GetPubSubSpec(repo repocpi.Repository) (pubsub.PubSubSpec, error) {
+	impl, err := repocpi.GetRepositoryImplementation(repo)
+	if err != nil {
+		return nil, err
+	}
+	gen, ok := impl.(*genericocireg.RepositoryImpl)
+	if !ok {
+		return nil, errors.ErrNotSupported("non-oci based ocm repository")
+	}
+
+	ocirepo := path.Join(gen.Meta().SubPath, componentmapping.ComponentDescriptorNamespace)
+	acc, err := gen.OCIRepository().LookupArtifact(ocirepo, META)
+	if errors.IsErrNotFound(err) || errors.IsErrUnknown(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot access meta data manifest version")
+	}
+	defer acc.Close()
```

Comment:
are we sure acc cannot be "open" if lookup artifact returns an error?

Reply:
if an error is returned acc is always nil

Reply:
s.a.